rust/worker/src/synchronizer.rs

```diff
@@ -0,0 +1,225 @@
+                            .expect("Failed to measure time")
+                            .as_millis();
+
+                        let mut missing = Vec::new();
+                        for digest in digests {
+                            // Ensure we do not send twice the same sync request.
+                            if self.pending.contains_key(&digest) {
+                                continue;
+                            }
+
+                            // Check if we received the batch in the meantime.
+                            match self.store.read(digest.to_vec()).await {
+                                Ok(None) => {
+                                    missing.push(digest.clone());
+                                    debug!("Requesting sync for batch {}", digest);
+                                },
+                                Ok(Some(_)) => {
+                                    // The batch arrived in the meantime: no need to request it.
+                                },
+                                Err(e) => {
+                                    error!("{}", e);
+                                    continue;
+                                }
+                            }
+
+                            // Add the digest to the waiter.
+                            let deliver = digest.clone();
+                            let (tx_cancel, rx_cancel) = channel(1);
+                            let fut = Self::waiter(digest.clone(), self.store.clone(), deliver, rx_cancel);
+                            waiting.push(fut);
+                            self.pending.insert(digest, (self.round, tx_cancel, now));
+                        }
+
+                        // Send sync request to a single node. If this fails, we will send it
+                        // to other nodes when a timer times out.
+                        let address = match self.committee.worker(&target, &self.id) {
+                            Ok(address) => address.worker_to_worker,
+                            Err(e) => {
+                                error!("The primary asked us to sync with an unknown node: {}", e);
+                                continue;
+                            }
+                        };
+                        let message = WorkerMessage::BatchRequest(missing, self.name);
+                        let serialized = bincode::serialize(&message).expect("Failed to serialize our own message");
+                        self.network.send(address, Bytes::from(serialized)).await;
+                    },
+                    PrimaryWorkerMessage::Cleanup(round) => {
+                        // Keep track of the primary's round number.
+                        self.round = round;
+
+                        // Cleanup internal state.
+                        if self.round < self.gc_depth {
+                            continue;
+                        }
+
+                        let mut gc_round = self.round - self.gc_depth;
```

Comment:
why is this mut?

Reply:
the function `retain` requires a mutable reference as input. We do not mutate gc_round. Is there a way to write this better?